### PR TITLE
Feature/rd++

### PR DESCRIPTION
**Pull request by cjy513203427**
## 📝 Description

- Provide a clear summary of the changes and the issue that has been addressed.
- 🛠️ Fixes # (issue number)

## ✨ Changes

Select what type of change your PR is:

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔨 Refactor (non-breaking change which refactors the code base)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔒 Security update

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [ x ] 📋 I have summarized my changes in the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) and followed the guidelines for my type of change (skip for minor changes, documentation updates, and test enhancements).
- [ ] 📚 I have made the necessary updates to the documentation (if applicable).
- [ x ] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).

For more information about code review checklists, see the [Code Review Checklist](https://github.com/openvinotoolkit/anomalib/blob/main/docs/source/markdown/guides/developer/code_review_checklist.md).

I re-implemented part of the RD++ algorithm. It is based on reverse distillation. I added `self-supervised optimal transport loss`, `reconstruction loss`, `contrast loss` and `Multiscale projection layers` based on📄 [Paper](https://openaccess.thecvf.com/content/CVPR2023/papers/Tien_Revisiting_Reverse_Distillation_for_Anomaly_Detection_CVPR_2023_paper.pdf)
🧑‍💻 [Code](https://github.com/tientrandinh/Revisiting-Reverse-Distillation)

Some things are missing: The authors use a customised dataloader and [noise.py](https://github.com/tientrandinh/Revisiting-Reverse-Distillation/blob/main/dataset/noise.py) for MVTEC dataset. However, I don't find any noise definition in Anomalib. Perlin noise is something else. Should I add a noise function in mvtec.py which could affect other functions or create a customised dataloader for RD++?

### CODE REVIEW

**Comment by cjy513203427** [comment]
This is test code (I used the recommended epochs from the authors.):
```python
import logging
from anomalib import TaskType
from anomalib.data import MVTec
from anomalib.engine import Engine
from anomalib.models import RevisitingReverseDistillation
from lightning.pytorch.callbacks import EarlyStopping, ModelCheckpoint

# configure logger
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

# Define the number of epochs for each category
epoch_mapping = {
    'carpet': 10,
    'leather': 10,
    'grid': 260,
    'tile': 260,
    'wood': 100,
    'cable': 240,
    'capsule': 300,
    'hazelnut': 160,
    'metal_nut': 160,
    'screw': 280,
    'toothbrush': 280,
    'transistor': 300,
    'zipper': 300,
    'pill': 200,
    'bottle': 200,
}

# datasets = ['screw', 'pill', 'capsule', 'carpet', 'grid', 'tile', 'wood', 'zipper', 'cable', 'toothbrush', 'transistor',
#             'metal_nut', 'bottle', 'hazelnut', 'leather']
# datasets = ['carpet']
datasets = ['bottle', 'hazelnut', 'leather']

for dataset in datasets:
    logger.info(f"================== Processing dataset: {dataset} ==================")
    task = TaskType.SEGMENTATION
    datamodule = MVTec(
        root="../datasets/MVTec",
        category=dataset,
        image_size=256,
        train_batch_size=32,
        eval_batch_size=32,
        num_workers=0,
        task=task,
    )

    model = RevisitingReverseDistillation()

    callbacks = [
        ModelCheckpoint(
            mode="max",
            monitor="pixel_AUROC",
        ),
        EarlyStopping(
            monitor="pixel_AUROC",
            mode="max",
            patience=3,
        ),
    ]

    # Get the number of epochs for the current dataset
    num_epochs = epoch_mapping.get(dataset, 100)  # Default to 100 if not found

    logger.info(f"Using {num_epochs} epochs for dataset: {dataset}")

    engine = Engine(
        max_epochs=num_epochs,
        check_val_every_n_epoch=3,
        callbacks=callbacks,
        pixel_metrics=["AUROC", "PRO"], image_metrics=["AUROC", "PRO"],
        accelerator="auto",  # <"cpu", "gpu", "tpu", "ipu", "hpu", "auto">,
        devices=1,
        logger=False,
    )

    logger.info(f"================== Start training for dataset: {dataset} ==================")
    engine.fit(datamodule=datamodule, model=model)

    logger.info(f"================== Start testing for dataset: {dataset} ==================")
    engine.test(datamodule=datamodule, model=model)
```

This is the result compared to `RD`:
|           | Screw          | Pill           | Capsule        | Carpet          | Grid           | Tile            | Wood           | Zipper         | Cable          | Toothbrush      | Transistor      | Metal Nut       | Bottle          | Hazelnut        | Leather         | Average        | Source         |
|-----------|----------------|----------------|----------------|-----------------|----------------|-----------------|----------------|----------------|----------------|-----------------|-----------------|-----------------|-----------------|-----------------|-----------------|----------------|----------------|
| RD        | **98.03**       | **97.63**       | **97.93**       | **99.36**        | **95.49**       | **100.00**      | **99.39**       | **97.16**       | **95.45**       | 91.39           | **97.87**        | **100.00**      | **100.00**      | **100.00**      | **100.00**      | **97.98**       | **98.50**      |
| RD++      | 94.71              | 94.44              | 90.65              | **99.80**               | 90.98              | **99.96**               | **99.39**              | 85.95              | 94.79              | 92.22               | **97.29**               | **99.22**               | **100**               | **100**               |  **100**               |  **100**              |  **100**              |

**Comment by samet-akcay** [comment]
@cjy513203427 thanks a lot for your contribution! 

regarding the noise, it could be a torchvision transform such as `SimplexNoise`, and could be added as a `train_transform` from `DataModule` or from the model. Since this is a model specific transform, I think it might make more sense to define this as a model specific transform such as done here:
https://github.com/openvinotoolkit/anomalib/blob/c00e101f3351b622e407dc097135fe3348860a9a/src/anomalib/models/image/patchcore/lightning_model.py#L131-L143

**Comment by haimat** [comment]
Hey guys, do you have any ETA on this?

**Comment by haimat** [comment]
Hi, has this been added to anomalib now?
